translator/trace/protospan_to_jaegerthrift.go

```diff
@@ -0,0 +1,381 @@
+	if ocSpanLinks == nil || ocSpanLinks.Link == nil {
+		return nil, nil
+	}
+
+	ocLinks := ocSpanLinks.Link
+	jRefs := make([]*jaeger.SpanRef, 0, len(ocLinks))
+	for _, ocLink := range ocLinks {
+		traceIDLow, traceIDHigh, err := traceIDBytesToLowAndHigh(ocLink.TraceId)
+		if err != nil {
+			return nil, fmt.Errorf("OC link has invalid trace ID: %v", err)
+		}
+
+		var jRefType jaeger.SpanRefType
+		switch ocLink.Type {
+		case tracepb.Span_Link_PARENT_LINKED_SPAN:
+			jRefType = jaeger.SpanRefType_CHILD_OF
+		default:
+			// TODO: (@pjanotti) Jaeger doesn't have a unknown SpanRefType, it has FOLLOWS_FROM or CHILD_OF
+			// at first mapping all others to FOLLOWS_FROM.
+			jRefType = jaeger.SpanRefType_FOLLOWS_FROM
+		}
+
+		spanID, err := ocIDBytesToJaegerID(ocLink.SpanId)
+		if err != nil {
+			return nil, fmt.Errorf("OC link has invalid span ID: %v", err)
+		}
+
+		jRef := &jaeger.SpanRef{
+			TraceIdLow:  traceIDLow,
+			TraceIdHigh: traceIDHigh,
+			RefType:     jRefType,
+			SpanId:      spanID,
+		}
+		jRefs = append(jRefs, jRef)
+	}
+
+	return jRefs, nil
+}
+
+func appendJaegerTagFromOCSpanKind(jTags []*jaeger.Tag, ocSpanKind tracepb.Span_SpanKind) []*jaeger.Tag {
+	// We could check if the key is already present but it doesn't seem worth at this point.
+	// TODO: (@pjanotti): Replace any OpenTracing literals by importing github.com/opentracing/opentracing-go/ext?
+	var tagValue string
+	switch ocSpanKind {
+	case tracepb.Span_CLIENT:
+		tagValue = "client"
+	case tracepb.Span_SERVER:
+		tagValue = "server"
+	}
+
+	if tagValue != "" {
+		jTag := &jaeger.Tag{
+			Key:  "span.kind",
+			VStr: &tagValue,
+		}
+		jTags = append(jTags, jTag)
+	}
+
+	return jTags
+}
+
+func ocTimeEventsToJaegerLogs(ocSpanTimeEvents *tracepb.Span_TimeEvents) []*jaeger.Log {
+	if ocSpanTimeEvents == nil || ocSpanTimeEvents.TimeEvent == nil {
+		return nil
+	}
+
+	ocTimeEvents := ocSpanTimeEvents.TimeEvent
+
+	// Assume that in general no time events are going to produce nil Jaeger logs.
+	jLogs := make([]*jaeger.Log, 0, len(ocTimeEvents))
+	for _, ocTimeEvent := range ocTimeEvents {
+		jLog := &jaeger.Log{
+			Timestamp: timestampToEpochMicroseconds(ocTimeEvent.Time),
+		}
+		switch teValue := ocTimeEvent.Value.(type) {
+		case *tracepb.Span_TimeEvent_Annotation_:
+			jLog.Fields = ocAnnotationToJagerTags(teValue.Annotation)
+		case *tracepb.Span_TimeEvent_MessageEvent_:
+			jLog.Fields = ocMessageEventToJaegerTags(teValue.MessageEvent)
+		default:
+			msg := "An unknown OpenCensus TimeEvent type was detected when translating to Jaeger"
+			jTag := &jaeger.Tag{
+				Key:  "unknown.oc.timeevent.type",
+				VStr: &msg,
+			}
+			jLog.Fields = append(jLog.Fields, jTag)
+		}
+
+		jLogs = append(jLogs, jLog)
+	}
+
+	return jLogs
+}
+
+func ocAnnotationToJagerTags(annotation *tracepb.Span_TimeEvent_Annotation) []*jaeger.Tag {
+	if annotation == nil {
+		return nil
+	}
+
+	// TODO: (@pjanotti) what about Description? Does it fit as another tag?
```

Comment:
Please see my comment about MessageEventToJaegerTags being "Span Logs" for Jaeger